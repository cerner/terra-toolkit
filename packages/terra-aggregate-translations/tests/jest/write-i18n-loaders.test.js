/* globals spyOn */
const fse = require('fs-extra');
const path = require('path');
// eslint-disable-next-line import/no-extraneous-dependencies
const MemoryFileSystem = require('memory-fs');
const writeI18nLoaders = require('../../src/write-i18n-loaders');

global.console = { warn: jest.fn() };

const locales = ['en', 'en-US', 'es'];
const memoryFS = new MemoryFileSystem();
const testFileSystems = { fse, memoryFS };

Object.keys(testFileSystems).forEach((testFS) => {
  describe(`write compiled i18n loaders for ${testFS} fileSystem`, () => {
    const fileSystem = testFileSystems[testFS];
    const outputDir = '/aggregated-translations';
    let writtenFilePaths;
    let writtenFiles;
    beforeEach(() => {
      writtenFilePaths = [];
      writtenFiles = [];
      spyOn(fileSystem, 'writeFileSync').and.callFake((fileName, hash) => {
        writtenFilePaths.push(fileName);
        writtenFiles.push(hash);
      });
      // eslint-disable-next-line no-console
      console.warn.mockClear();
    });

    it('writes compiled loader files', () => {
      const outputFiles = [
        path.resolve(process.cwd(), outputDir, 'intlLoaders.js'),
        path.resolve(process.cwd(), outputDir, 'translationsLoaders.js'),
      ];

      writeI18nLoaders(locales, fileSystem, outputDir);

      expect(writtenFilePaths).toEqual(outputFiles);
      expect(writtenFiles[0]).toMatchSnapshot();
      expect(writtenFiles[1]).toMatchSnapshot();
      expect(writtenFiles[2]).toMatchSnapshot();
    });

    it('writes compiled loader files with modern format if specified', () => {
      const outputFiles = [
        path.resolve(process.cwd(), outputDir, 'intlLoaders.js'),
        path.resolve(process.cwd(), outputDir, 'translationsLoaders.js'),
      ];

      writeI18nLoaders(locales, fileSystem, outputDir, 'es6');

      expect(writtenFilePaths).toEqual(outputFiles);
      expect(writtenFiles[0]).toMatchSnapshot();
      expect(writtenFiles[1]).toMatchSnapshot();
      expect(writtenFiles[2]).toMatchSnapshot();
    });

    it('logs a warning message if a locale is not a terra-supported locale', () => {
      writeI18nLoaders(['cy'], fileSystem, outputDir);

      /* eslint-disable no-console */
      expect(console.warn).toHaveBeenCalledTimes(1);
      expect(console.warn).toBeCalledWith(expect.stringContaining('WARNING: cy is NOT a Terra supported locale. Creating a translation and intl loader for cy, but'));
      /* eslint-enable no-console */
    });

    it('writes a translation file for a non-terra-supported locale', () => {
      writeI18nLoaders(['cy'], fileSystem, outputDir);

      expect(writtenFiles[0]).toMatchSnapshot();
      expect(writtenFiles[1]).toMatchSnapshot();
      expect(writtenFiles[2]).toMatchSnapshot();
    });
  });
});
