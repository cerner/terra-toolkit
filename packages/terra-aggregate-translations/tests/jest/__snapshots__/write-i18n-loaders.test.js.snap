// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`write compiled i18n loaders for fse fileSystem writes a translation file for a non-terra-supported locale 1`] = `
"'use strict';

var loadCyIntl = function loadCyIntl() {
  return require.ensure([], function (require) {
    return require('intl/locale-data/jsonp/cy.js');
  }, 'cy-intl-local');
};

var intlLoaders = {
  'cy': loadCyIntl
};

module.exports = intlLoaders;"
`;

exports[`write compiled i18n loaders for fse fileSystem writes a translation file for a non-terra-supported locale 2`] = `
"'use strict';

var loadCyTranslations = function loadCyTranslations(callback, scope) {
  return require.ensure([], function (require) {
    // eslint-disable-next-line
    var i18n = require('./cy.js');
    callback.call(scope, i18n);
    return i18n;
  }, 'cy-translations');
};

var translationsLoaders = {
  'cy': loadCyTranslations
};

module.exports = translationsLoaders;"
`;

exports[`write compiled i18n loaders for fse fileSystem writes a translation file for a non-terra-supported locale 3`] = `undefined`;

exports[`write compiled i18n loaders for fse fileSystem writes compiled loader files 1`] = `
"'use strict';

var loadEnIntl = function loadEnIntl() {
  return require.ensure([], function (require) {
    return require('intl/locale-data/jsonp/en.js');
  }, 'en-intl-local');
};

var loadEnUSIntl = function loadEnUSIntl() {
  return require.ensure([], function (require) {
    return require('intl/locale-data/jsonp/en-US.js');
  }, 'en-US-intl-local');
};

var loadEsIntl = function loadEsIntl() {
  return require.ensure([], function (require) {
    return require('intl/locale-data/jsonp/es.js');
  }, 'es-intl-local');
};

var intlLoaders = {
  'en': loadEnIntl,
  'en-US': loadEnUSIntl,
  'es': loadEsIntl
};

module.exports = intlLoaders;"
`;

exports[`write compiled i18n loaders for fse fileSystem writes compiled loader files 2`] = `
"'use strict';

var loadEnTranslations = function loadEnTranslations(callback, scope) {
  return require.ensure([], function (require) {
    // eslint-disable-next-line
    var i18n = require('./en.js');
    callback.call(scope, i18n);
    return i18n;
  }, 'en-translations');
};

var loadEnUSTranslations = function loadEnUSTranslations(callback, scope) {
  return require.ensure([], function (require) {
    // eslint-disable-next-line
    var i18n = require('./en-US.js');
    callback.call(scope, i18n);
    return i18n;
  }, 'en-US-translations');
};

var loadEsTranslations = function loadEsTranslations(callback, scope) {
  return require.ensure([], function (require) {
    // eslint-disable-next-line
    var i18n = require('./es.js');
    callback.call(scope, i18n);
    return i18n;
  }, 'es-translations');
};

var translationsLoaders = {
  'en': loadEnTranslations,
  'en-US': loadEnUSTranslations,
  'es': loadEsTranslations
};

module.exports = translationsLoaders;"
`;

exports[`write compiled i18n loaders for fse fileSystem writes compiled loader files 3`] = `undefined`;

exports[`write compiled i18n loaders for fse fileSystem writes compiled loader files with modern format if specified 1`] = `
"'use strict';

const loadEnIntl = () =>
   import('intl/locale-data/jsonp/en.js')
   .catch(error => console.log('An error occurred while loading intl/locale-data/jsonp/en.js' + \\"\\\\n\\" + error));

const loadEnUSIntl = () =>
   import('intl/locale-data/jsonp/en-US.js')
   .catch(error => console.log('An error occurred while loading intl/locale-data/jsonp/en-US.js' + \\"\\\\n\\" + error));

const loadEsIntl = () =>
   import('intl/locale-data/jsonp/es.js')
   .catch(error => console.log('An error occurred while loading intl/locale-data/jsonp/es.js' + \\"\\\\n\\" + error));

var intlLoaders = {
  'en': loadEnIntl,
  'en-US': loadEnUSIntl,
  'es': loadEsIntl
};

module.exports = intlLoaders;"
`;

exports[`write compiled i18n loaders for fse fileSystem writes compiled loader files with modern format if specified 2`] = `
"'use strict';

const loadEnTranslations = (callback, scope) =>
   import( /* webpackChunkName: \\"en-translations\\" */ 'en.js')
     .then((module) => { callback.call(scope, module);})
     .catch(error => console.log('An error occurred while loading en.js' + \\"\\\\n\\" + error));

const loadEnUSTranslations = (callback, scope) =>
   import( /* webpackChunkName: \\"en-US-translations\\" */ 'en-US.js')
     .then((module) => { callback.call(scope, module);})
     .catch(error => console.log('An error occurred while loading en-US.js' + \\"\\\\n\\" + error));

const loadEsTranslations = (callback, scope) =>
   import( /* webpackChunkName: \\"es-translations\\" */ 'es.js')
     .then((module) => { callback.call(scope, module);})
     .catch(error => console.log('An error occurred while loading es.js' + \\"\\\\n\\" + error));

var translationsLoaders = {
  'en': loadEnTranslations,
  'en-US': loadEnUSTranslations,
  'es': loadEsTranslations
};

module.exports = translationsLoaders;"
`;

exports[`write compiled i18n loaders for fse fileSystem writes compiled loader files with modern format if specified 3`] = `undefined`;

exports[`write compiled i18n loaders for memoryFS fileSystem writes a translation file for a non-terra-supported locale 1`] = `
"'use strict';

var loadCyIntl = function loadCyIntl() {
  return require.ensure([], function (require) {
    return require('intl/locale-data/jsonp/cy.js');
  }, 'cy-intl-local');
};

var intlLoaders = {
  'cy': loadCyIntl
};

module.exports = intlLoaders;"
`;

exports[`write compiled i18n loaders for memoryFS fileSystem writes a translation file for a non-terra-supported locale 2`] = `
"'use strict';

var loadCyTranslations = function loadCyTranslations(callback, scope) {
  return require.ensure([], function (require) {
    // eslint-disable-next-line
    var i18n = require('./cy.js');
    callback.call(scope, i18n);
    return i18n;
  }, 'cy-translations');
};

var translationsLoaders = {
  'cy': loadCyTranslations
};

module.exports = translationsLoaders;"
`;

exports[`write compiled i18n loaders for memoryFS fileSystem writes a translation file for a non-terra-supported locale 3`] = `undefined`;

exports[`write compiled i18n loaders for memoryFS fileSystem writes compiled loader files 1`] = `
"'use strict';

var loadEnIntl = function loadEnIntl() {
  return require.ensure([], function (require) {
    return require('intl/locale-data/jsonp/en.js');
  }, 'en-intl-local');
};

var loadEnUSIntl = function loadEnUSIntl() {
  return require.ensure([], function (require) {
    return require('intl/locale-data/jsonp/en-US.js');
  }, 'en-US-intl-local');
};

var loadEsIntl = function loadEsIntl() {
  return require.ensure([], function (require) {
    return require('intl/locale-data/jsonp/es.js');
  }, 'es-intl-local');
};

var intlLoaders = {
  'en': loadEnIntl,
  'en-US': loadEnUSIntl,
  'es': loadEsIntl
};

module.exports = intlLoaders;"
`;

exports[`write compiled i18n loaders for memoryFS fileSystem writes compiled loader files 2`] = `
"'use strict';

var loadEnTranslations = function loadEnTranslations(callback, scope) {
  return require.ensure([], function (require) {
    // eslint-disable-next-line
    var i18n = require('./en.js');
    callback.call(scope, i18n);
    return i18n;
  }, 'en-translations');
};

var loadEnUSTranslations = function loadEnUSTranslations(callback, scope) {
  return require.ensure([], function (require) {
    // eslint-disable-next-line
    var i18n = require('./en-US.js');
    callback.call(scope, i18n);
    return i18n;
  }, 'en-US-translations');
};

var loadEsTranslations = function loadEsTranslations(callback, scope) {
  return require.ensure([], function (require) {
    // eslint-disable-next-line
    var i18n = require('./es.js');
    callback.call(scope, i18n);
    return i18n;
  }, 'es-translations');
};

var translationsLoaders = {
  'en': loadEnTranslations,
  'en-US': loadEnUSTranslations,
  'es': loadEsTranslations
};

module.exports = translationsLoaders;"
`;

exports[`write compiled i18n loaders for memoryFS fileSystem writes compiled loader files 3`] = `undefined`;

exports[`write compiled i18n loaders for memoryFS fileSystem writes compiled loader files with modern format if specified 1`] = `
"'use strict';

const loadEnIntl = () =>
   import('intl/locale-data/jsonp/en.js')
   .catch(error => console.log('An error occurred while loading intl/locale-data/jsonp/en.js' + \\"\\\\n\\" + error));

const loadEnUSIntl = () =>
   import('intl/locale-data/jsonp/en-US.js')
   .catch(error => console.log('An error occurred while loading intl/locale-data/jsonp/en-US.js' + \\"\\\\n\\" + error));

const loadEsIntl = () =>
   import('intl/locale-data/jsonp/es.js')
   .catch(error => console.log('An error occurred while loading intl/locale-data/jsonp/es.js' + \\"\\\\n\\" + error));

var intlLoaders = {
  'en': loadEnIntl,
  'en-US': loadEnUSIntl,
  'es': loadEsIntl
};

module.exports = intlLoaders;"
`;

exports[`write compiled i18n loaders for memoryFS fileSystem writes compiled loader files with modern format if specified 2`] = `
"'use strict';

const loadEnTranslations = (callback, scope) =>
   import( /* webpackChunkName: \\"en-translations\\" */ 'en.js')
     .then((module) => { callback.call(scope, module);})
     .catch(error => console.log('An error occurred while loading en.js' + \\"\\\\n\\" + error));

const loadEnUSTranslations = (callback, scope) =>
   import( /* webpackChunkName: \\"en-US-translations\\" */ 'en-US.js')
     .then((module) => { callback.call(scope, module);})
     .catch(error => console.log('An error occurred while loading en-US.js' + \\"\\\\n\\" + error));

const loadEsTranslations = (callback, scope) =>
   import( /* webpackChunkName: \\"es-translations\\" */ 'es.js')
     .then((module) => { callback.call(scope, module);})
     .catch(error => console.log('An error occurred while loading es.js' + \\"\\\\n\\" + error));

var translationsLoaders = {
  'en': loadEnTranslations,
  'en-US': loadEnUSTranslations,
  'es': loadEsTranslations
};

module.exports = translationsLoaders;"
`;

exports[`write compiled i18n loaders for memoryFS fileSystem writes compiled loader files with modern format if specified 3`] = `undefined`;
